File: recreate_flatten.py
#!/usr/bin/env python

# represent desired output
def strhack(arr):
  hack = str(arr).replace(" [],","").replace("[", "").replace("]", "")
  return [int(x) for x in hack.split(", ")]

def flatten(arr):
  flat = arr[:] # clone it
  i = 0
  while (i < len(flat)):
    if (type(flat[i]) is list):
      flat[i] = flat[i][:] # clone it
      if len(flat[i]) == 0:
        del flat[i]
      else:
        tmp = flat[i]
        flat[i] = tmp.pop(0)
        if len(tmp) >  0:
          flat.insert(i + 1, tmp)
    else:
      i +=1
  return flat

input = [1,2,3,4,[2,34],[],[23,[232,23,[1]]]]

print input
print strhack(input)
print input
print flatten(input)
print input

"""

[1, 2, 3, 4, [2, 34], [], [23, [232, 23, [1]]]]
[1, 2, 3, 4, 2, 34, 23, 232, 23, 1]
[1, 2, 3, 4, [2, 34], [], [23, [232, 23, [1]]]]
[1, 2, 3, 4, 2, 34, 23, 232, 23, 1]
[1, 2, 3, 4, [2, 34], [], [23, [232, 23, [1]]]]

"""

File: recreate_flatten.rb
#!/usr/bin/env ruby

class Array
  def go_flat
    flat = self.clone
    i = 0
    while (i < flat.length)
      if flat[i].is_a?(Array)
        if flat[i].empty?
          flat.delete_at(i)
        else
          tmp = flat[i].clone
          flat[i] = tmp.shift
          flat.insert(i + 1, tmp) unless tmp.empty?
        end
      else
        i +=1
      end
    end
    flat
  end
end

input = [1,2,3,4,[2,34],[],[23,[232,23,[1]]]]

puts input.inspect
puts input.flatten.inspect
puts input.inspect
puts input.go_flat.inspect
puts input.inspect

__END__

[1, 2, 3, 4, [2, 34], [23, [232, 23, [1]]]]
[1, 2, 3, 4, 2, 34, 23, 232, 23, 1]
[1, 2, 3, 4, [2, 34], [23, [232, 23, [1]]]]
[1, 2, 3, 4, 2, 34, 23, 232, 23, 1]
[1, 2, 3, 4, [2, 34], [23, [232, 23, [1]]]]
